pkg/core/diamond_commit.go

```diff
@@ -0,0 +1,326 @@
+package core
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	iradix "github.com/hashicorp/go-immutable-radix"
+	"github.com/oneconcern/datamon/pkg/core/status"
+	"github.com/oneconcern/datamon/pkg/errors"
+	"github.com/oneconcern/datamon/pkg/model"
+	"github.com/oneconcern/datamon/pkg/storage"
+	"github.com/segmentio/ksuid"
+	"go.uber.org/zap"
+)
+
+// Commit a diamond
+func (d *Diamond) Commit(opts ...Option) error {
+	return d.implCommit(opts...)
+}
+
+// Cancel a diamond
+func (d *Diamond) Cancel() error {
+	// TODO(fred)
+	return status.ErrNotImplemented.WrapMessage("diamond cancel")
+}
+
+// WithState returns the diamond with its state modified and transition timestamp updated
+func (d *Diamond) WithState(state model.DiamondState) *Diamond {
+	d.DiamondDescriptor.State = state
+	d.DiamondDescriptor.StateChangeTime = time.Now()
```

Comment:
realizing i might be entering into the discussion too late on this implementation detail since i wasn't present for all of the design phase:

is there any way we can avoid storing times from the machines where the datamon binary runs and instead use times provided by the GCS service?  that way, we'll have the exceedingly nice property of synchronization.  [`ObjectAttrs`](https://godoc.org/cloud.google.com/go/storage#ObjectAttrs) keeps track of times.  i'd be happier with a patch that tears the `storage.Store` up a little (as we're already doing with conditionals CRC) in order to avoid unsynchronized timestamps appearing in the data sets that it's this program's responsibility to organize and track.

Reply:
I understand the sync issue, but I don't know of any such service specifically for GCP. You might otherwise subscribe to time.google.com as your NTP server. I am guessing that's how our pods are subscribed to.

Reply:
> subscribe to time.google.com as your NTP server. I am guessing that's how our pods

to iron out the user story somewhat, can we _either_ make sure

* pods are time-synced (@bluelion ?)

and

* users know (via the markdown in this readme, confluence, or just direct correspondence) that the `diamond` and `split` functionality is mainly intended for in-k8s programs, and we should be notified in #datamon or otherwise if `diamond` and `split` are used from local machines, at least for data sets that might find their way into production

_or_

* include a ± time constant to encode the uncertainty, used along with (or to derive) `diamondCommitLockExpiry` and any other future compairsions with `StateChangeTime`?

Reply:
indeed you have a point here - I don't have a good solution for it right now. Mentioned it in the detailed design document. See https://github.com/oneconcern/datamon/pull/399